File: leetcode/medium/3-longestSubstringWithoutRepeatingCharacters.js
// Given a string s, find the length of the longest substring without repeating characters.

// Example 1:

// Input: s = "abcabcbb"
// Output: 3
// Explanation: The answer is "abc", with the length of 3.

/**
 * @param {string} s
 * @return {number}
 */
 
// first pass solution
let lengthOfLongestSubstring = function(s) {;
    let currLongest = [];
    let maxLongest = [];
    let left = 0;
    let right = 0;
    
    if (s.length <= 1) return s.length
                                            
    while(left < s.length && right < s.length) {
        if (!currLongest.includes(s[right])) {
            currLongest.push(s[right]);
            right++;
        }
        
        if (currLongest.length > maxLongest.length) {
            maxLongest = [...currLongest];
        }
        
        if (currLongest.includes(s[right])){
            currLongest.shift();
            left++;
        }
    }                     
    return maxLongest.length;
};

// second pass solution
var lengthOfLongestSubstring = function(s) {
    let set = new Set();
    let left = 0;
    let right = 0;
    let max = 0;
    
    while (left < s.length && right < s.length) {
        if (!set.has(s[right])) {
            set.add(s[right]);
            right++;
        } else {
            set.delete(s[left]);
            left++;
        }
        max = Math.max(set.size, max);
    }
    return max;
}

File: leetcode/medium/3-longestSubstringWithoutRepeatingCharacters.java
// first pass solution
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int output = 0;
        
        for (int i = 0; i < s.length(); i++) {
            Set<Character> set = new HashSet<>();
            
            for (int j = i; j < s.length(); j++) {
                if (set.contains(s.charAt(j))) {
                    break;
                }
                set.add(s.charAt(j));
                output = Math.max(output, set.size());
            }
        }
        return output;
    }
}

// second pass solution
class Solution {
    public int lengthOfLongestSubstring(String s) {
        Set<Character> set = new HashSet<>();
        int output = 0;
        int left = 0;
        int right = 0;
        
        while (right < s.length()) {
            if (!set.contains(s.charAt(right))) {
                set.add(s.charAt(right));
                right++;
            } else {
                set.remove(s.charAt(left));
                left++;
            }
            output = Math.max(output, set.size());
        }
        return output;
    }
}
